### ISSUE TITLE
Mobi Links / TOC not implemented

### ISSUE THREAD

**Issue by Poiol000** [koreader/koreader#11621]
* KOReader version:The newly released
* Device:
kindle kpw4 5.16.2.1.1
#### Issue
why after everything is configured, books in koreader can not use the table of contents to jump, nor can they display references by clicking on links, which can be done for books outside the library?

**Comment by Frenzie** [comment]
Could you try to explain a bit more what you're doing and what you're expecting? If you tap on an entry in the table of contents normally it goes to that page.

**Comment by Poiol000** [comment]
1. I tap on the contents of the book in Koreader that I expects to jump to the target page, but it doesn't work(It turns page instead of jumping to the target goal).
2. When there is a comment in the book, I click to turn the page and do not pop up the comment (after according to the online tutorial set up the link jump).

**Comment by jonnyl2** [comment]
For 1., it sounds like you're expecting KOReader to follow a book-internal link when tapping on it, but KOReader just turns the page?

Make sure the setting `Tap to follow links` is on under the following menu:

- Top menu -> Gear icon (3rd menu) -> Taps and gestures -> Links -> 

For 2., I'm not sure if you _want_ the footnote to pop up, or you _don't_ want it to pop up. If you want the footnote to pop up when you tap a reference link, the above setting will also fix that.

**Comment by Poiol000** [comment]
Thanks for your help!
Maybe I have find out the cause that the **mobi** format cant support links jump. When I use **epub** format books, it works correctly.

Wish to fix this problem(**mobi's jump**) in the future.

**Comment by Frenzie** [comment]
Mobi support is rudimentary and unlikely to improve, but I can probably confirm with the Alice in Wonderland document from https://filesamples.com/formats/mobi 

The very last page seems to have links, but they use `filepos` instead of `href`. If there's an `id=etc` somewhere maybe something can be done with that. Pinging @poire-z on the off chance he knows anything about this without diving into the code.

<img src=https://github.com/koreader/koreader/assets/202757/73992572-b38a-40df-b8e9-a898010dd7e8 width=45%>

**Comment by poire-z** [comment]
Not knowing anything, not wanting to know :)
Not caring about mobi, caring about FB2 is already a lot :)

**Comment by Frenzie** [comment]
Haha, fair enough.

**Comment by Frenzie** [comment]
I became briefly curious about it, but it's too much trouble for the gains because it's too far removed from regular HTML. For reference, here's part of what calibre does.
https://github.com/kovidgoyal/calibre/blame/b5c688c94deff46965676f0b872f0a929d572154/src/calibre/ebooks/mobi/reader/mobi6.py#L852-L880

I didn't bother trying to fully comprehend it, but I think it's finding a byte position in the file and inserting an `id="filepos-1234"` at that location, and replacing the `filepos=1234` by `href="#filepos-1234"`. For someone with more motivation this kind of preprocessing might be relatively simple to add though, in the initial opening phase prior to XML parsing.

There is however a simple solution:
```
mobitool -e Alices\ Adventures\ in\ Wonderland.mobi
```
(rename to .epub)
[Alices Adventures in Wonderland.epub.zip](https://github.com/koreader/koreader/files/14894719/Alices.Adventures.in.Wonderland.epub.zip)

<hr>

Incidentally, the ToC seems to be referenced at the beginning as follows:
```
<guide><reference type="toc" title="Table of Contents" filepos=0000199447 /></guide>
```
calibre doesn't do anything with that.

**Comment by Frenzie** [comment]
@BenRobart
> Ah, eesh... would it not be a relatively simple fix? I was hoping to not need to convert my whole library to .epub!

Replying here to keep it all together.

---

Conceptually it's quite simple, but practically not so much.

In some sense it's much easier to get something that works if you can clearly define the problem and a desired solution than it was two years ago.

<details>
<img width="1852" height="1022" alt="Image" src="https://github.com/user-attachments/assets/fbfad8ce-9b9d-4163-a470-fe0cbe4a0b1b" />
</details>

[mobi-link.patch](https://github.com/user-attachments/files/27130320/mobi-link.patch)

I'm not enthused by the overhead of an entire minimal HTML parser (and obviously MOBI isn't designed to be parsed that way) but I don't really see another way to do it. But this method is also rather poor and if I hadn't been able to do it by listing the requirements to a parrot in a minute or two I'd never have bothered even making that into a POC…

**Comment by BenRobart** [comment]
> [@BenRobart](https://github.com/BenRobart)
> 
> > Ah, eesh... would it not be a relatively simple fix? I was hoping to not need to convert my whole library to .epub!
> 
> Replying here to keep it all together.
> 
> Conceptually it's quite simple, but practically not so much.
> 
> In some sense it's much easier to get something that works if you can clearly define the problem and a desired solution than it was two years ago.
> Details
> <img alt="Image" width="1852" height="1022" src="https://private-user-images.githubusercontent.com/202757/584259125-fbfad8ce-9b9d-4163-a470-fe0cbe4a0b1b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzczMDIzNDUsIm5iZiI6MTc3NzMwMjA0NSwicGF0aCI6Ii8yMDI3NTcvNTg0MjU5MTI1LWZiZmFkOGNlLTliOWQtNDE2My1hNDcwLWZlMGNiZTRhMGIxYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNDI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDQyN1QxNTAwNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NWEyOTFjZjY4MjNjNDRiMWM3NDZmNGU3ZGU3NjE3ZDc5ZTVjNTQ2MmYxOTdkZjViOTI1MDlkYTJmZDg0MjViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.ZNpKWhplOTRaT4n0D-_iPaVsYKN0Isi1wEbDt-zbTLg">
> 
> [mobi-link.patch](https://github.com/user-attachments/files/27130320/mobi-link.patch)
> 
> I'm not enthused by the overhead of an entire minimal HTML parser (and obviously MOBI isn't designed to be parsed that way) but I don't really see another way to do it. But this method is also rather poor and if I hadn't been able to do it by listing the requirements to a parrot in a minute or two I'd never have bothered even making that into a POC…

Thank you for sending this over, I'll give it a go! Do I just put this in the patches folder and it's good to go?

**Comment by Frenzie** [comment]
No, this is a source code patch for crengine. It requires compilation. It's just a follow-up to my earlier comment <https://github.com/koreader/koreader/issues/11621#issuecomment-2041171621> that describes the strategy by showing an actual working example of the strategy.

**Comment by poire-z** [comment]
Spent more time reading and trying to understand that patch - than you did copiloting it :)
A bit misleaded by some of the explanation in your black screenshot (ie. `6 Place only marker in the preprocessed HTML` which I thought was clever - but they end up also in the DOM)
I guess you asked it to do as calibre ? There are lots of clever things.
(And some bits I would change, ie the _mobiIdNodeMap, that it doesn't serialize in the cache (and rebuild after each opening from cache), unlike our _idNodeMap which serialized, dunno if it's on purpose or it just didn't see it, or just didn't want to introduce too much changes.)
(It's again always a bit unconfortable reviewing some AI output when I can't ask it questions :) or suggest things and see what it thinks.)

If filepos points to the middle of a text node, this can happen:
<img width="425" height="116" alt="Image" src="https://github.com/user-attachments/assets/f16a8a26-e4a5-47c1-8dd0-0a5268dfc352" />
so, possibly broken kerning, which is minor - but past highlights (on a single text node) possibly broken as this would generate 2 next nodes, and a highlight spanning over that A would give:
```
  pos0 = "/html[1]/body/body/p[6]/text()[4].205",
  pos1 = "/html[1]/body/body/p[6]/text()[5].13",
```
so would need a bump to DOM_VERSION_CURRENT and this implemented conditionally so it's not triggered on past open books and we don't break highlights.

I wonder if in the wild, filepos always point to a real element, and never to the middle of a text node (I'm not sure how a tool could produce such a filepos without some upstream refernce to an existing element.
If we could assume that, and expect filepos is inside some tag, and if not we could use the previous tag.
If that was possible, we could put filepos-id in the preprocessed HTML, and when parsing that in the 2nd phase, transform them into a real id= (if there's no other), or use an existing id when we rewrite the filepos to a href=. That way, less noise, no _idNodeMap, and the changes would be really limited to pdbfmt.cpp.

(Would also need to add the attibute names in fb2def.h so they get consistant internal ids, to avoid some kind of unrealted problems.)

**Comment by Frenzie** [comment]
> I guess you asked it to do as calibre ? There are lots of clever things.

You can crudely summarize it that way, yes, but I doubt you'd get a working result if you merely told it to do as calibre. ;-)

The numbered list I shared was its plan based on my instructions and a tiny bit of back and forth before starting actual implementation.

**Comment by GammaWolf** [comment]
I also tried KOReader and fiddled with the settings for quite some time before seeing that  TOC wasn't working for any of the books put on the kindle touch with calibre. With such basic functionality not working, sadly it's not usable for me. Since there's a wontfix tagged, maybe it would be good to add a disclamer text near the kindle download that this doesn't work for this case and people needing that don't spend their time needlessly. There's a hint somewhere that mobi support is limited, but I wouldn't have guessed that the TOC doesn't work. Otherwise, it's great.

**Comment by BenRobart** [comment]
> I also tried KOReader and fiddled with the settings for quite some time before seeing that TOC wasn't working for any of the books put on the kindle touch with calibre. With such basic functionality not working, sadly it's not usable for me. Since there's a wontfix tagged, maybe it would be good to add a disclamer text near the kindle download that this doesn't work for this case and people needing that don't spend their time needlessly. There's a hint somewhere that mobi support is limited, but I wouldn't have guessed that the TOC doesn't work. Otherwise, it's great.

With how often it's reported and how easy the fix seemingly is, it's somewhat shocking that nothing's been done about it. For me, I'll probably just switch to another reader app.
